<a href="https://colab.research.google.com/github/SallyPeter/gomycodeDSbootcamp/blob/main/Python/Checkpoint_Data_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Instructions**

1. Load the dataset, display the ten first lines, store the results in a variable called 'client_0_bills'.
2. What is the data type of the 'client_0_bills' variable ?
3. Display the general information of the dataset and try to answer the following questions :
4. How many rows and columns do we have in this dataset ?
5. How many categorical features are present in the dataset ?
6. How much memory space does the dataset consume ?
7. Inspect the dataset for potential missing values.
8. Select your strategy to handle missing values, and tell us why you had made that choice.
9. Run a descriptive analysis on numeric features (columns).
Select the bills records for the client with an id ='train_Client_0', using 2 methods.
10. Transform the 'counter_type' feature to a numeric variable using the encoder of your choice.
11. Delete the 'counter_statue' feature from the Dataframe

*This code uses the data STEG_BILLING_HISTORY.csv*

In [1]:
import pandas as pd

In [2]:
# Load the dataset, display the ten first lines, store the results in a variable called 'client_0_bills'.
data = pd.read_csv('STEG_BILLING_HISTORY.csv')

client_0_bills = data.head(10)


<ipython-input-2-cc4ed85e76b2>:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('STEG_BILLING_HISTORY.csv')


In [4]:
# What is the data type of the 'client_0_bills' variable ?
type(client_0_bills)

pandas.core.frame.DataFrame

It is a pandas DataFrame

In [5]:
# Display the general information of the dataset and try to answer the following questions :
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2615629 entries, 0 to 2615628
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   float64
 8   consommation_level_1  float64
 9   consommation_level_2  float64
 10  consommation_level_3  float64
 11  consommation_level_4  float64
 12  old_index             float64
 13  new_index             float64
 14  months_number         float64
 15  counter_type          object 
dtypes: float64(10), int64(2), object(4)
memory usage: 319.3+ MB


In [7]:
# How many rows and columns do we have in this dataset ?
data.shape

(2615629, 16)

Data has over 2.6 million rows of data and 16 columns

In [8]:
# How many categorical features are present in the dataset ?
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2615629 entries, 0 to 2615628
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   float64
 8   consommation_level_1  float64
 9   consommation_level_2  float64
 10  consommation_level_3  float64
 11  consommation_level_4  float64
 12  old_index             float64
 13  new_index             float64
 14  months_number         float64
 15  counter_type          object 
dtypes: float64(10), int64(2), object(4)
memory usage: 319.3+ MB


From the .info() method, data has 4 categorical features


### How much memory space does the dataset consume?

319.3MB

In [9]:
# Inspect the dataset for potential missing values.
data.isnull().sum()

client_id                  0
invoice_date               0
tarif_type                 0
counter_number            23
counter_statue             0
counter_code               0
reading_remarque        2611
counter_coefficient        1
consommation_level_1       1
consommation_level_2       1
consommation_level_3       1
consommation_level_4       1
old_index                  1
new_index                  1
months_number              1
counter_type               1
dtype: int64

The features with missing values are 'counter_number', 'reading_remarque', 'counter_coefficient', the consommation features, old and new index, 'months_number' and 'counter_type' with 'reading_remarque having the most missing values.

In [ ]:
# Select your strategy to handle missing values, and tell us why you had made that choice.


I choose to drop the missing values as 2611 is less than 1% of the records in the data. Droping it has liitle to no effect to the dataset

In [15]:
data = data.dropna(subset=['reading_remarque', 'counter_number', 'counter_coefficient', 'consommation_level_1', 'consommation_level_2', 'consommation_level_3', 'consommation_level_4',
                                                        'old_index', 'new_index', 'months_number', 'counter_type'])

data.isnull().sum()




client_id               0
invoice_date            0
tarif_type              0
counter_number          0
counter_statue          0
counter_code            0
reading_remarque        0
counter_coefficient     0
consommation_level_1    0
consommation_level_2    0
consommation_level_3    0
consommation_level_4    0
old_index               0
new_index               0
months_number           0
counter_type            0
dtype: int64

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2612994 entries, 0 to 2615627
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   float64
 8   consommation_level_1  float64
 9   consommation_level_2  float64
 10  consommation_level_3  float64
 11  consommation_level_4  float64
 12  old_index             float64
 13  new_index             float64
 14  months_number         float64
 15  counter_type          object 
dtypes: float64(10), int64(2), object(4)
memory usage: 338.9+ MB


In [17]:
# Run a descriptive analysis on numeric features (columns).
num_col = data.select_dtypes(include='number')

num_col.describe()

tarif_type  counter_number  counter_code  reading_remarque  \
count  2.612994e+06    2.612994e+06  2.612994e+06      2.612994e+06   
mean   2.009691e+01    1.220926e+11  1.729026e+02      7.320786e+00   
std    1.346245e+01    1.651865e+12  1.339192e+02      1.641789e+00   
min    8.000000e+00    0.000000e+00  0.000000e+00      6.000000e+00   
25%    1.100000e+01    1.211750e+05  5.000000e+00      6.000000e+00   
50%    1.100000e+01    4.949180e+05  2.030000e+02      8.000000e+00   
75%    4.000000e+01    1.115336e+06  2.070000e+02      9.000000e+00   
max    4.500000e+01    2.798115e+13  6.000000e+02      4.130000e+02   

       counter_coefficient  consommation_level_1  consommation_level_2  \
count         2.612994e+06          2.612994e+06          2.612994e+06   
mean          1.003320e+00          4.114430e+02          1.089441e+02   
std           3.434935e-01          8.523503e+02          1.043448e+03   
min           0.000000e+00          0.000000e+00          0.000000e+00   
25%           1.000000e+00          8.000000e+01          0.000000e+00   
50%           1.000000e+00          2.740000e+02          0.000000e+00   
75%           1.000000e+00          6.010000e+02          0.000000e+00   
max           5.000000e+01          9.999100e+05          8.198860e+05   

       consommation_level_3  consommation_level_4     old_index     new_index  \
count          2.612994e+06          2.612994e+06  2.612994e+06  2.612994e+06   
mean           2.056795e+01          5.276381e+01  1.782218e+04  1.841513e+04   
std            1.709811e+02          8.557036e+02  4.000883e+04  4.061404e+04   
min            0.000000e+00          0.000000e+00  0.000000e+00  0.000000e+00   
25%            0.000000e+00          0.000000e+00  1.802000e+03  2.070000e+03   
50%            0.000000e+00          0.000000e+00  7.737000e+03  8.240000e+03   
75%            0.000000e+00          0.000000e+00  2.176600e+04  2.246200e+04   
max            6.449200e+04          3.435680e+05  2.800280e+06  2.870972e+06   

       months_number  
count   2.612994e+06  
mean    5.670593e+01  
std     3.496651e+03  
min     1.000000e+00  
25%     4.000000e+00  
50%     4.000000e+00  
75%     4.000000e+00  
max     6.366240e+05

In [18]:
data.columns

Index(['client_id', 'invoice_date', 'tarif_type', 'counter_number',
       'counter_statue', 'counter_code', 'reading_remarque',
       'counter_coefficient', 'consommation_level_1', 'consommation_level_2',
       'consommation_level_3', 'consommation_level_4', 'old_index',
       'new_index', 'months_number', 'counter_type'],
      dtype='object')

In [19]:
# Select the bills records for the client with an id ='train_Client_0', using 2 methods.

# Method 1

data[data.client_id == 'train_Client_0']

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0            203               8.0                  1.0                  82.0   
1            203               6.0                  1.0                1200.0   
2            203               8.0                  1.0                 123.0   
3            207               8.0                  1.0                 102.0   
4            207               9.0                  1.0                 572.0   
5            207               9.0                  1.0                 314.0   
6            207               9.0                  1.0                 541.0   
7            207               9.0                  1.0                 585.0   
8            203               9.0                  1.0                1200.0   
9            203               6.0                  1.0                1082.0   
10           203               6.0                  1.0                 174.0   
11           203               6.0                  1.0                 628.0   
12           203               6.0                  1.0                 886.0   
13           203               8.0                  1.0           

In [22]:
# Method 2

data.loc[data.client_id == 'train_Client_0']

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0            203               8.0                  1.0                  82.0   
1            203               6.0                  1.0                1200.0   
2            203               8.0                  1.0                 123.0   
3            207               8.0                  1.0                 102.0   
4            207               9.0                  1.0                 572.0   
5            207               9.0                  1.0                 314.0   
6            207               9.0                  1.0                 541.0   
7            207               9.0                  1.0                 585.0   
8            203               9.0                  1.0                1200.0   
9            203               6.0                  1.0                1082.0   
10           203               6.0                  1.0                 174.0   
11           203               6.0                  1.0                 628.0   
12           203               6.0                  1.0                 886.0   
13           203               8.0                  1.0           

In [23]:
data.head()

client_id invoice_date  tarif_type  counter_number counter_statue  \
0  train_Client_0   2014-03-24          11       1335667.0              0   
1  train_Client_0   2013-03-29          11       1335667.0              0   
2  train_Client_0   2015-03-23          11       1335667.0              0   
3  train_Client_0   2015-07-13          11       1335667.0              0   
4  train_Client_0   2016-11-17          11       1335667.0              0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203               8.0                  1.0                  82.0   
1           203               6.0                  1.0                1200.0   
2           203               8.0                  1.0                 123.0   
3           207               8.0                  1.0                 102.0   
4           207               9.0                  1.0                 572.0   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                   0.0                   0.0                   0.0   
1                 184.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   old_index  new_index  months_number counter_type  
0    14302.0    14384.0            4.0         ELEC  
1    12294.0    13678.0            4.0         ELEC  
2    14624.0    14747.0            4.0         ELEC  
3    14747.0    14849.0            4.0         ELEC  
4    15066.0    15638.0           12.0         ELEC

In [24]:
data.counter_type.unique()

array(['ELEC', 'GAZ'], dtype=object)

In [25]:
from sklearn.preprocessing import LabelEncoder

In [26]:
# Transform the 'counter_type' feature to a numeric variable using the encoder of your choice.
le = LabelEncoder()

data.counter_type = le.fit_transform(data.counter_type)

data.head()

client_id invoice_date  tarif_type  counter_number counter_statue  \
0  train_Client_0   2014-03-24          11       1335667.0              0   
1  train_Client_0   2013-03-29          11       1335667.0              0   
2  train_Client_0   2015-03-23          11       1335667.0              0   
3  train_Client_0   2015-07-13          11       1335667.0              0   
4  train_Client_0   2016-11-17          11       1335667.0              0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203               8.0                  1.0                  82.0   
1           203               6.0                  1.0                1200.0   
2           203               8.0                  1.0                 123.0   
3           207               8.0                  1.0                 102.0   
4           207               9.0                  1.0                 572.0   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                   0.0                   0.0                   0.0   
1                 184.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   old_index  new_index  months_number  counter_type  
0    14302.0    14384.0            4.0             0  
1    12294.0    13678.0            4.0             0  
2    14624.0    14747.0            4.0             0  
3    14747.0    14849.0            4.0             0  
4    15066.0    15638.0           12.0             0

In [27]:
# Delete the 'counter_statue' feature from the Dataframe

data.drop('counter_statue', axis=1, inplace=True)

data.head()

client_id invoice_date  tarif_type  counter_number  counter_code  \
0  train_Client_0   2014-03-24          11       1335667.0           203   
1  train_Client_0   2013-03-29          11       1335667.0           203   
2  train_Client_0   2015-03-23          11       1335667.0           203   
3  train_Client_0   2015-07-13          11       1335667.0           207   
4  train_Client_0   2016-11-17          11       1335667.0           207   

   reading_remarque  counter_coefficient  consommation_level_1  \
0               8.0                  1.0                  82.0   
1               6.0                  1.0                1200.0   
2               8.0                  1.0                 123.0   
3               8.0                  1.0                 102.0   
4               9.0                  1.0                 572.0   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                   0.0                   0.0                   0.0   
1                 184.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   old_index  new_index  months_number  counter_type  
0    14302.0    14384.0            4.0             0  
1    12294.0    13678.0            4.0             0  
2    14624.0    14747.0            4.0             0  
3    14747.0    14849.0            4.0             0  
4    15066.0    15638.0           12.0             0